### DISCUSSION TITLE
Is there an example on how to use mosaicing?

### DISCUSSION THREAD

**Discussion by Nakilon** [libvips/ruby-vips#287] [Q&A] [answered]
https://libvips.github.io/libvips/API/current/libvips-mosaicing.html

I have a thousand of images that all have different amount of matching so I probably can't come up with universal numeric arguments.for vips_match. Also I can't tell I understand them at all. When I had a look on docs previously I believe I saw a function that produces a 2d bw map of level of matching between two images so I could then decide what do to with it. But I probably can't find what function that was and the...

```none
irb(main):005:0> im11.match(im12)
Vips::Error: unable to call match: you supplied 2 arguments, but operation needs 10.
```

just tells me I need something different.

UPD: something like this: https://stackoverflow.com/q/36688439/322020
I believe I saw such function...

Or maybe that was `match` but I imagines it to work simplier when I was reading docs the last time.

UPD2: Or I guess it was `spcor`. After `scaleimage` (btw, is it per band?) gives me such image:

![image](https://user-images.githubusercontent.com/2870363/110770619-93a2db00-826a-11eb-9bed-b6ae7ebd062f.png)

but takes eternity to correlate two 1280x720 images while I have a thousand of them...

UPD3: oh, and I just realised that it's only trying one quarter of relative shifting so I need to do it 4 times, right?

```ruby
rd = img11res.fastcor(img12res)
ru = img11res.flipver.fastcor(img12res.flipver).flipver
ld = img11res.fliphor.fastcor(img12res.fliphor).fliphor
lu = img11res.fliphor.flipver.fastcor(img12res.fliphor.flipver).fliphor.flipver
Vips::Image.arrayjoin([lu,ru,ld,rd], across:2).scaleimage.write_to_file "temp.png"
```
![image](https://user-images.githubusercontent.com/2870363/110777214-ee8c0080-8271-11eb-9dc9-ef387834ef9a.png)
does not look right...

Maybe I don't understand how it's supposed to work. It makes no sense to me.
```ruby
a = Vips::Image.new_from_array [[0,0,0,0,0],[0,1,0,0,0],[0,0,0,0,0],[0,0,0,0,0]]
```
```ruby
b = Vips::Image.new_from_array [[0,0,0,0,0],[0,0,0,0,0],[0,0,0,1,0],[0,0,0,0,0]]
```
![image](https://user-images.githubusercontent.com/2870363/110785775-62cba180-827c-11eb-837a-7ada434916d4.png)
![image](https://user-images.githubusercontent.com/2870363/110785821-6eb76380-827c-11eb-96b1-84915a11384e.png)
```ruby
Vips::Image.arrayjoin([lu,ru,ld,rd], across:2).scaleimage.write_to_file "temp.png"
```
![image](https://user-images.githubusercontent.com/2870363/110785605-3021a900-827c-11eb-8e73-b2406dc7bf55.png)

**Comment by jcupitt** [comment]
Yes, `spcor` is very slow -- it computes the full normalized spatial correlation, so 1.0 is perfect correlation and 0.0 is perfect mismatch.

There's `vips_fastcor()` -- it just computes sum of squares of differences, so you get 0 for a perfect match and some huge number for a perfect mismatch.

**Reply by Nakilon** [reply]
Yep, I tried both. But I don't understand why those 4 results don't assemble smoothly. And absolutely don't understand the last example with two 5x4 images that should match when shifted by (2,1).

**Comment by Nakilon** [comment, marked as answer]
I guess I tried the wrong function. `phasecor` seems to be what I need:
```none
 img12res.phasecor(img11res).stats.to_a.drop(1).transpose[8,2]
=> [[[0.0], [0.0], [0.0]], [[23.0], [23.0], [23.0]]]
```
These as I understand are the shifting candidates per band.

What I don't yet understand is why it's `#<Image 160x90 double, 3 bands, b-w>` -- being b-w but with 3 bands makes the `.write_to_file` save both dark and light pixels as white. ~~How do I switch the interpretation?~~ `.copy(interpretation: :srgb)`

P.S.: and it's weird that when I call it on two similar (i.e. no shift needed) or even identical images it's all a noise and no bright pixel in the left-top corner.
P.P.S: also there is a true point and three false points and they are I don't know how to chose one because they all are of the same intensity. For example, here the true one is the left bottom: 
![image](https://user-images.githubusercontent.com/2870363/111031425-95aa9c80-8418-11eb-8ba1-046e45ec45f2.png)

**Reply by Nakilon** [reply]
So I decided to just chose the one of 4 shifts by the lowest `(rect1-rect2).abs.avg`.
Then I used `embed` and `bandrank` and got this:
![image](https://user-images.githubusercontent.com/2870363/111037664-33609480-8436-11eb-8045-3fa9ddbfa5bb.png)
The new problem is that taking "median" among several images using `bandrank` may take the black color while I would like to discard it. Like if here https://www.rubydoc.info/gems/ruby-vips/Vips/BlendMode was such blendmode as "discard black and pick the one closest to the average". Otherwise the median results in this for now:
![image](https://user-images.githubusercontent.com/2870363/111038474-39587480-843a-11eb-8e63-60ff66ff7770.png)
Or if I chose the `index:images.size-1` it results in a mess:
![image](https://user-images.githubusercontent.com/2870363/111038910-61e16e00-843c-11eb-96ba-8f9915526b90.png)

**Reply by Nakilon** [reply]
To illustrate the result of "discard black and pick the one closest to the average", here is a rough implementation on Ruby level (that is obviously ~100 times slower i.e. unusable):
![image](https://user-images.githubusercontent.com/2870363/111043398-5f3e4300-8453-11eb-8495-a459bc3c0781.png)
